pyro/distributions/empirical.py

```diff
@@ -155,21 +95,20 @@ def mean(self):
 
     @property
     def variance(self):
-        self._finalize()
         if self._samples.dtype in (torch.int32, torch.int64):
             raise ValueError("Variance for discrete empirical distribution undefined. " +
                              "Consider converting samples to ``torch.float32`` " +
                              "or ``torch.float64``. If these are samples from a " +
                              "`Categorical` distribution, consider converting to a " +
                              "`OneHotCategorical` distribution.")
-        deviation_squared = torch.pow(self._samples - self.mean, 2)
+        mean = self.mean.unsqueeze(self._aggregation_dim)
+        deviation_squared = torch.pow(self._samples - mean, 2)
         return self._weighted_mean(deviation_squared)
 
-    def get_samples_and_weights(self):
-        self._finalize()
-        return self._samples, self._log_weights
+    @property
+    def weights(self):
+        return self._log_weights
```

Comment:
I think we need to do some math here to get the weight. Otherwise, we might rename `weights` to `log_weights`.

Reply:
I think we can rename this to `log_weights`, since `Empirical` only accepts `log_weights`. I'll change to use `log_weights` everywhere, we don't need to make the interface confusing when the user can use a one liner to transform their weights.